PHOENIX-3744: Support snapshot scanners for MR-based Non-aggregate queries

akshita-malhotra opened the pull request:
- Parallel Scan grouper is extended to differentiate the functionality for getting region boundaries

- Added integration test, compares the snapshot read result with the result from select query by setting CurrentScn value.

- the configuration parameter is the snapshot name key, if set do a snapshot read

- Used an existing PhoenixIndexDBWritable class for the purpose of testing, will add a new one as I will add more tests.

- ExpressionProjector functionality is extended for snapshots as the keyvalue format returned from TableSnapshotScanner is different from ClientScanner and therefore not properly interrupted by Phoenix thereby returning null in case of projected columns.
For the same table, following shows the different format of the keyvalues:

1. ClientScanner:
keyvalues={AAPL/_v:\x00\x00\x00\x01/1493061452132/Put/vlen=7/seqid=0/value=SSDD }

2. TableSnapshotScanner:
keyvalues={AAPL/0:\x00\x00\x00\x00/1493061673859/Put/vlen=1/seqid=4/value=x, 
AAPL/0:\x80\x0B/1493061673859/Put/vlen=4/seqid=4/value=SSDD}

@JamesRTaylor @lhofhansl 

To DO:
Add more integration tests to cover different scenarios such as where clause etc

akshita-malhotra commented:
- Snapshot scanner for non-aggregate queries.
- Added integration tests (simple select query, conditional and limit)
- Abstracted out ScanRegionObserver code to fetch the processed region scanner without the coprocessor environment
- To make the snapshot work for aggregate queries, due to the complexity of the aggregate region observer  code it is almost impossible to refactor the code without fully understanding the functionality. Will require some guidance if the aggregate query use case is required.
@JamesRTaylor @lhofhansl

JamesRTaylor commented:
I don't think it's necessary to fully understand the functionality to do the refactoring I've mentioned, @akshita-malhotra. Here's how I'd recommend approaching it:

* create a new interface solely for the purpose of only abstracting RegionCoprocessorEnvironment access called RegionContext. The interface would have at least two methods: getRegion and getConfiguration. We might need more if other methods are called in RegionCoprocessorEnvironment.
* have two implementations of this interface: RegionCoprocessorContext and RegionShapshotContext. The constructor of RegionCoprocessorContext would take a RegionCoprocessorEnvironment as an argument, while the RegionShapshotContext would take a Region and Configuration.
* do an across the board replace of RegionCoprocessorEnvironment with RegionContext. You can likely not do this for secondary index related code (org.apache.phoenix.hbase.index.Indexer and PhoenixTransactionalIndexer). You'll find out here if other methods are called from RegionCoprocessorEnvironment or ObserverContext (which can be dealt with in a variety of ways, for example by throwing an UnsupportedOperationException if need be in the snapshot implementation).
* in the top level coprocessor methods that pass in RegionCoprocessorEnvironment (mostly abstract BaseScannerRegionObserver class), instantiate a RegionCoprocessorContext by passing in the RegionCoprocessorEnvironment. From this point onward, all access will go through the RegionContext interface.

You could do this refactoring completely separate from the PHOENIX-3744 so that you don't mix the two. Then PHOENIX-3744 would have something like a RegionScannerFactory (your RegionObserverUtil) that gives you back a RegionScanner given a RegionContext and you'd create a RegionShapshotContext as the backing implementation in your snapshot reading code.

JamesRTaylor commented:
Please amend your commit message to be prefixed with PHOENIX-3744 (instead of Phoenix-3744) so that review comments show up as comments on the JIRA, @akshita-malhotra. Also, let's get a patch attached to the JIRA so we can get a test run to make sure there are no regressions. Directions here: https://phoenix.apache.org/contributing.html#Generate_a_patch

akshita-malhotra commented:
Sure, I will do that. Thanks @JamesRTaylor

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:
```diff
@@ -496,6 +496,17 @@ public BaseResultIterators(QueryPlan plan, Integer perScanLimit, Integer offset,
 
+    private List<HRegionLocation> getRegionBoundaries(ParallelScanGrouper scanGrouper)
+        throws SQLException{
+        if (scanGrouper instanceof DefaultParallelScanGrouper) {
```
These casts should not be necessary. Either add an init method to ParallelScanGrouper or call the init from the implementing class in getRegionBoundaries.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/mapreduce/PhoenixInputFormat.java:
```diff
@@ -185,7 +188,10 @@ private QueryPlan getQueryPlan(final JobContext context, final Configuration con
             }
             // Initialize the query plan so it sets up the parallel scans
-            queryPlan.iterator(MapReduceParallelScanGrouper.getInstance());
+            if(snapshotName != null)
```
Add an init method to the ParallelScanGrouper interface that returns an instance and have the TableMRParallelScanGrouper implementation of that return TableMRParallelScanGrouper.getInstance(). We shouldn't need switching logic here based on snapshotName.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/monitoring/MetricType.java:
```diff
@@ -25,6 +25,7 @@
     QUERY_TIME("Query times"),
     NUM_PARALLEL_SCANS("Number of scans that were executed in parallel"),
     SCAN_BYTES("Number of bytes read by scans"),
+    SNAPSHOT_SCAN_BYTES("Number of bytes read by snapshot scans"),
```
Do we need a different metric here? Or would it make more sense to use SCAN_BYTES for the snapshot case too?

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/util/NonAggregateROUtil.java:
```diff
@@ -0,0 +1,355 @@
+
+import static org.apache.phoenix.util.EncodedColumnsUtil.getMinMaxQualifiersFromScan;
+
+public class NonAggregateROUtil extends RegionObserverUtil {
```
In general, our Util classes only have static methods. How about a different naming convention for these that a post fixed with "Factory" instead of "Util, like NonAggregateRegionScannerFactory?

JamesRTaylor commented on phoenix-core/src/test/java/org/apache/phoenix/query/BaseTest.java:
```diff
@@ -592,7 +592,7 @@ private static String initMiniCluster(Configuration conf, ReadOnlyProps override
         utility = new HBaseTestingUtility(conf);
         try {
-            utility.startMiniCluster(NUM_SLAVES_BASE);
+            utility.startMiniCluster(NUM_SLAVES_BASE, true);
```
Just curious on this change. Is it necessary?

JamesRTaylor commented:
Thanks for the updates, @akshita-malhotra - I made a few comments/questions.

akshita-malhotra commented:
Thanks @JamesRTaylor . I squashed the commits and changed the prefix of the commit message. I will answer/make appropriate changes and upload the patch onto the jira.

akshita-malhotra commented:
@JamesRTaylor 
- Changed ParallelScanGrouper classes as per the review
- Changes to BaseTest were to avoid the following error:
"Restore directory cannot be a sub directory of HBase root directory"
Therefore, was sending true to create the root dir. Changed to use a random dir instead to avoid to make these changes
- Refactored the util classes to Factory

Also, uploaded the patch on the jira.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/ScanningResultIterator.java:
```diff
@@ -38,10 +38,12 @@
 public class ScanningResultIterator implements ResultIterator {
     private final ResultScanner scanner;
     private final CombinableMetric scanMetrics;
+    private final boolean snapshotScan;
```
Looks like the changes to this file aren't necessary. Please revert.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/PhoenixMapReduceUtil.java:
```diff
@@ -63,6 +67,66 @@ public static void setInput(final Job job, final Class<? extends DBWritable> inp
+     * @param conditions Condition clause to be added to the WHERE clause. Can be <tt>null</tt> if there are no conditions.
+     * @param fieldNames fields being projected for the SELECT query.
+     */
+    public static void setInput(final Job job, final Class<? extends DBWritable> inputClass, final String snapshotName, String tableName,
```
Why wouldn't we want to take the snapshot here instead of passing in the snapshot name?

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/DefaultParallelScanGrouper.java:
```diff
@@ -17,46 +17,79 @@
-            }
-        }
-        return startNewScanGroup;
+  private static DefaultParallelScanGrouper INSTANCE = new DefaultParallelScanGrouper();
```
I don't think that DefaultParallelScanGrouper can be a singleton with the state of context and tableName inside of it.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/TableMRParallelScanGrouper.java:
```diff
@@ -0,0 +1,99 @@
+
+public class TableMRParallelScanGrouper extends MapReduceParallelScanGrouper {
+
+  private static TableMRParallelScanGrouper INSTANCE = null;
```
Same comment here - how can this be a singleton?

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/mapreduce/PhoenixInputFormat.java:
```diff
@@ -185,7 +184,7 @@ private QueryPlan getQueryPlan(final JobContext context, final Configuration con
             }
             // Initialize the query plan so it sets up the parallel scans
-            queryPlan.iterator(MapReduceParallelScanGrouper.getInstance());
+            queryPlan.iterator(TableMRParallelScanGrouper.init(configuration));
```
Can't this just instantiate a TableMRParallelScanGrouper and the init method can be invoked in the constructor so that it doesn't need to be exposed in the interface?

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/mapreduce/util/PhoenixConfigurationUtil.java:
```diff
@@ -192,6 +196,18 @@ public static void setOutputTableName(final Configuration configuration, final S
+    public static void setSnapshotNameKey(final Configuration configuration, final String snapshotName) {
+        Preconditions.checkNotNull(configuration);
+        Preconditions.checkNotNull(snapshotName);
+        configuration.set(SNAPSHOT_NAME_KEY, snapshotName);
```
The idea of having the snapshot name in the configuration is not going to translate well when we want to expose snapshot reads for queries in general as the configuration is a global object.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/iterate/RegionContext.java:
```diff
@@ -0,0 +1,15 @@
+public interface RegionContext {
+
+  /** @return the Coprocessor environment */
+  RegionCoprocessorEnvironment getEnvironment();
```
If this is the only method, we don't really need this interface. Why not just keep using RegionCoprocessorEnvironment in our APIs and you can just instantiate you snapshot-based one prior to getting the RegionScanner?

JamesRTaylor commented:
Thanks for the revision, @akshita-malhotra. It's looking very good. I made some comments inline.

akshita-malhotra commented:
@JamesRTaylor  Thanks a lot for the review. I have made the suggested changes and uploaded the updated patch on the jira. 
Regarding creating snapshot to generalize the use of snapshots for M/R jobs, I was under the impression that we are passing the snapshot name as input after our last discussion with @lhofhansl and Rahul G.
If we are to follow the former approach, I will go ahead and make the changes.

JamesRTaylor commented:
Patch looks very good, @akshita-malhotra. What's the advantage, @lhofhansl, of forcing users to create the snapshot themselves before starting the job? Wouldn't it be simpler for the snapshot to be created during the setup/initialization of the MR job? In the non MR case, when we want to support running arbitrary queries over snapshot(s), seems like we'd want Phoenix to create them, no? Otherwise, we'd need to provide the user with some means of associating a snapshot with a table name (which might get cumbersome). The alternative is to let Phoenix manage this transparently.

JamesRTaylor commented:
Had an offline discussion with @lhofhansl and he brought up the use case of having a snapshot at a known point in time (i.e. midnight last night) and wanting to run Phoenix queries over it through MR. That seems like a good reason to pass in the snapshot name rather than build it when the MR starts.

I'll commit this into 4.x and master. Nice work, @akshita-malhotra!

stoty commented:
Already merged.
